File: apps/desktop/scripts/test-gemini.mjs
#!/usr/bin/env node

import fs from "node:fs";
import path from "node:path";

const DEFAULT_BASE = "https://generativelanguage.googleapis.com/v1beta";
const DEFAULT_MODEL = "gemini-3.1-flash-lite-preview";

function loadLocalEnv() {
  const envPath = path.resolve(process.cwd(), ".env.local");
  if (!fs.existsSync(envPath)) return false;
  const text = fs.readFileSync(envPath, "utf8");
  for (const line of text.split(/\r?\n/)) {
    const trimmed = line.trim();
    if (!trimmed || trimmed.startsWith("#")) continue;
    const eq = trimmed.indexOf("=");
    if (eq <= 0) continue;
    const key = trimmed.slice(0, eq).trim();
    let value = trimmed.slice(eq + 1).trim();
    if ((value.startsWith('"') && value.endsWith('"')) || (value.startsWith("'") && value.endsWith("'"))) {
      value = value.slice(1, -1);
    }
    if (!process.env[key]) process.env[key] = value;
  }
  return true;
}

function argValue(name, fallback = "") {
  const prefix = `${name}=`;
  const hit = process.argv.slice(2).find((arg) => arg.startsWith(prefix));
  return hit ? hit.slice(prefix.length) : fallback;
}

function redact(value) {
  if (!value) return "";
  if (value.length <= 10) return `${value.slice(0, 2)}...`;
  return `${value.slice(0, 6)}...${value.slice(-4)}`;
}

function usage() {
  console.log(`Usage:
  GEMINI_API_KEY=your_key node scripts/test-gemini.mjs
  npm run test:gemini

Options:
  --model=gemini-3.1-flash-lite-preview
  --base=https://generativelanguage.googleapis.com/v1beta
  --timeout=30000
  --prompt="你好，返回一句话"
`);
}

const loadedEnv = loadLocalEnv();
const apiKey = process.env.GEMINI_API_KEY || process.env.GOOGLE_API_KEY || "";
const model = argValue("--model", process.env.GEMINI_MODEL || DEFAULT_MODEL).trim();
const base = argValue("--base", process.env.GEMINI_API_BASE || DEFAULT_BASE).trim().replace(/\/+$/, "");
const timeoutMs = Number(argValue("--timeout", process.env.GEMINI_TIMEOUT_MS || "30000"));
const prompt = argValue("--prompt", "请只回复一个 JSON 对象：{\"ok\":true,\"message\":\"pong\"}");

if (!apiKey.trim()) {
  console.error("Missing GEMINI_API_KEY. Example:");
  usage();
  process.exit(2);
}

if (!Number.isFinite(timeoutMs) || timeoutMs <= 0) {
  console.error(`Invalid timeout: ${timeoutMs}`);
  process.exit(2);
}

const url = `${base}/models/${encodeURIComponent(model)}:generateContent?key=${encodeURIComponent(apiKey.trim())}`;
const redactedUrl = `${base}/models/${encodeURIComponent(model)}:generateContent?key=${redact(apiKey.trim())}`;
const body = {
  contents: [{
    role: "user",
    parts: [{ text: prompt }],
  }],
  generationConfig: {
    temperature: 0.1,
  },
};

const startedAt = performance.now();
const controller = new AbortController();
const timer = setTimeout(() => controller.abort(), timeoutMs);

console.log("Gemini smoke test");
console.log(`Loaded .env.local: ${loadedEnv ? "yes" : "no"}`);
console.log(`URL: ${redactedUrl}`);
console.log(`Model: ${model}`);
console.log(`Timeout: ${timeoutMs}ms`);
console.log(`Prompt chars: ${prompt.length}`);

try {
  const response = await fetch(url, {
    method: "POST",
    headers: { "content-type": "application/json" },
    body: JSON.stringify(body),
    signal: controller.signal,
  });
  clearTimeout(timer);

  const elapsedMs = Math.round(performance.now() - startedAt);
  const text = await response.text();
  let json = null;
  try {
    json = JSON.parse(text);
  } catch {}

  console.log(`HTTP: ${response.status} ${response.statusText}`);
  console.log(`Elapsed: ${elapsedMs}ms`);

  if (!response.ok) {
    const message = json?.error?.message || text.slice(0, 1000);
    const status = json?.error?.status || "unknown";
    console.error(`Gemini error status: ${status}`);
    console.error(`Gemini error message: ${message}`);
    process.exit(1);
  }

  const output = json?.candidates?.[0]?.content?.parts
    ?.map((part) => part.text || "")
    .join("") || "";

  console.log(`Output chars: ${output.length}`);
  console.log("Output preview:");
  console.log(output.slice(0, 1200));
} catch (error) {
  clearTimeout(timer);
  const elapsedMs = Math.round(performance.now() - startedAt);
  console.error(`Request failed after ${elapsedMs}ms`);
  if (error?.name === "AbortError") {
    console.error(`Timeout after ${timeoutMs}ms`);
  } else {
    console.error(error);
  }
  process.exit(1);
}

File: apps/desktop/src-tauri/src/bin/test_gemini.rs
use serde_json::json;
use std::{env, fs, path::PathBuf, time::Instant};

const DEFAULT_BASE: &str = "https://generativelanguage.googleapis.com/v1beta";
const DEFAULT_MODEL: &str = "gemini-3.1-flash-lite-preview";

fn load_env_local() -> bool {
    let Ok(cwd) = env::current_dir() else {
        return false;
    };
    let candidates = [
        cwd.join("../.env.local"),
        cwd.join(".env.local"),
        cwd.join("../../.env.local"),
    ];
    let Some(path) = candidates.iter().find(|path| path.exists()) else {
        return false;
    };
    let Ok(text) = fs::read_to_string(path) else {
        return false;
    };
    for line in text.lines() {
        let trimmed = line.trim();
        if trimmed.is_empty() || trimmed.starts_with('#') {
            continue;
        }
        let Some((key, value)) = trimmed.split_once('=') else {
            continue;
        };
        if env::var_os(key.trim()).is_some() {
            continue;
        }
        let mut value = value.trim().to_string();
        if (value.starts_with('"') && value.ends_with('"'))
            || (value.starts_with('\'') && value.ends_with('\''))
        {
            value = value[1..value.len().saturating_sub(1)].to_string();
        }
        env::set_var(key.trim(), value);
    }
    true
}

fn arg_value(name: &str) -> Option<String> {
    let prefix = format!("{name}=");
    env::args()
        .skip(1)
        .find_map(|arg| arg.strip_prefix(&prefix).map(str::to_string))
}

fn redact(value: &str) -> String {
    if value.len() <= 10 {
        return format!("{}...", &value[..value.len().min(2)]);
    }
    format!("{}...{}", &value[..6], &value[value.len() - 4..])
}

fn format_reqwest_error(err: &reqwest::Error) -> String {
    let mut parts = vec![err.to_string()];
    let mut source = std::error::Error::source(err);
    while let Some(next) = source {
        parts.push(next.to_string());
        source = next.source();
    }
    redact_url_key(&parts.join(" | caused by: "))
}

fn redact_url_key(value: &str) -> String {
    let mut out = String::with_capacity(value.len());
    let mut rest = value;
    while let Some(idx) = rest.find("key=") {
        out.push_str(&rest[..idx + 4]);
        out.push_str("[REDACTED]");
        let tail = &rest[idx + 4..];
        let end = tail
            .find(|ch: char| ch == '&' || ch == ')' || ch.is_whitespace())
            .unwrap_or(tail.len());
        rest = &tail[end..];
    }
    out.push_str(rest);
    out
}

fn read_prompt() -> Result<String, String> {
    if let Some(path) = arg_value("--from-file") {
        let path = PathBuf::from(path);
        return fs::read_to_string(&path)
            .map_err(|err| format!("读取 prompt 文件失败 {:?}: {err}", path));
    }
    Ok(arg_value("--prompt")
        .or_else(|| env::var("GEMINI_PROMPT").ok())
        .unwrap_or_else(|| {
            "请只回复一个 JSON 对象：{\"ok\":true,\"message\":\"pong\"}".to_string()
        }))
}

#[tokio::main]
async fn main() {
    let loaded_env = load_env_local();
    let api_key = env::var("GEMINI_API_KEY")
        .or_else(|_| env::var("GOOGLE_API_KEY"))
        .unwrap_or_default();
    if api_key.trim().is_empty() {
        eprintln!("Missing GEMINI_API_KEY. Put it in apps/desktop/.env.local or export it.");
        std::process::exit(2);
    }

    let model = arg_value("--model")
        .or_else(|| env::var("GEMINI_MODEL").ok())
        .unwrap_or_else(|| DEFAULT_MODEL.to_string());
    let base = arg_value("--base")
        .or_else(|| env::var("GEMINI_API_BASE").ok())
        .unwrap_or_else(|| DEFAULT_BASE.to_string())
        .trim_end_matches('/')
        .to_string();
    let prompt = match read_prompt() {
        Ok(prompt) => prompt,
        Err(err) => {
            eprintln!("{err}");
            std::process::exit(2);
        }
    };
    let repeat = arg_value("--repeat")
        .or_else(|| env::var("GEMINI_REPEAT").ok())
        .and_then(|value| value.parse::<usize>().ok())
        .unwrap_or(1)
        .max(1);
    let url = format!(
        "{base}/models/{model}:generateContent?key={}",
        api_key.trim()
    );
    let body = json!({
        "contents": [{
            "role": "user",
            "parts": [{ "text": prompt }]
        }],
        "generationConfig": {
            "temperature": 0.1
        }
    });
    let body_bytes = match serde_json::to_vec(&body) {
        Ok(bytes) => bytes,
        Err(err) => {
            eprintln!("Serialize request failed: {err}");
            std::process::exit(2);
        }
    };

    println!("Gemini Rust reqwest smoke test");
    println!(
        "Loaded .env.local: {}",
        if loaded_env { "yes" } else { "no" }
    );
    println!(
        "URL: {base}/models/{model}:generateContent?key={}",
        redact(api_key.trim())
    );
    println!("Repeat: {repeat}");
    println!("Prompt chars: {}", prompt.chars().count());
    println!("Body bytes: {}", body_bytes.len());

    let client = reqwest::Client::new();
    let mut failures = 0usize;
    for i in 1..=repeat {
        let started = Instant::now();
        println!("--- request {i}/{repeat} ---");
        let response = match client
            .post(&url)
            .header(reqwest::header::CONTENT_TYPE, "application/json")
            .body(body_bytes.clone())
            .send()
            .await
        {
            Ok(response) => response,
            Err(err) => {
                failures += 1;
                eprintln!("Request failed after {}ms", started.elapsed().as_millis());
                eprintln!("{}", format_reqwest_error(&err));
                continue;
            }
        };

        let status = response.status();
        let text = response.text().await.unwrap_or_default();
        println!("HTTP: {status}");
        println!("Elapsed: {}ms", started.elapsed().as_millis());
        if !status.is_success() {
            failures += 1;
            eprintln!("Body preview:");
            eprintln!("{}", &text[..text.len().min(1200)]);
            continue;
        }
        println!("Body preview:");
        println!("{}", &text[..text.len().min(1200)]);
    }

    if failures > 0 {
        eprintln!("{failures}/{repeat} request(s) failed");
        std::process::exit(1);
    }
}
